File: 1382-balance-a-binary-search-tree/1382-balance-a-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def balanceBST(self, root: TreeNode) -> TreeNode:
        if not root:
            return None
        
        # Step 1: Create the backbone (vine)
        # Temporary dummy node
        vine_head = TreeNode(0)
        vine_head.right = root
        curr = vine_head
        while curr.right:
            if curr.right.left:
                self.right_rotate(curr, curr.right)
            else:
                curr = curr.right
                
        # Step 2: Count the nodes
        node_count = 0
        curr = vine_head.right
        while curr:
            node_count += 1
            curr = curr.right
            
        # Step 3: Create a balanced BST
        m = 2 ** int(log2(node_count + 1)) - 1
        self.make_rotations(vine_head, node_count - m)
        while m > 1:
            m >>= 1
            self.make_rotations(vine_head, m)
            
        balanced_root = vine_head.right
        # Delete the temporary dummy node
        vine_head = None
        return balanced_root
    
    # Function to perform a right rotation    
    def right_rotate(self, parent: TreeNode, node: TreeNode) -> None:
        tmp = node.left
        node.left = tmp.right
        tmp.right = node
        parent.right = tmp
        
    # Function to perform a left rotation
    def left_rotate(self, parent: TreeNode, node: TreeNode) -> None:
        tmp = node.right
        node.right = tmp.left
        tmp.left = node
        parent.right = tmp
        
    # Function to perform a series of left rotations to balance the vine
    def make_rotations(self, vine_head: TreeNode, count: int) -> None:
        curr = vine_head
        for _ in range(count):
            tmp = curr.right
            self.left_rotate(curr, tmp)
            curr = curr.right
    

File: 1382-balance-a-binary-search-tree/1382-balance-a-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    void inorder(TreeNode* node, vector<int>& traversal){
        // Perform an inorder traversal to store the elements in sorted order
        if (!node) return;
        inorder(node -> left, traversal);
        traversal.push_back(node -> val);
        inorder(node -> right, traversal);
    }
    
    TreeNode* createBalancedTree(vector<int>& inorderTraversal, int start, int end){
        // Base case: If the start index is greater than the end index, return
        // nullptr
        if (start > end) return nullptr;
        
        // Find the middle element of the current range
        int mid = start + ((end - start) >> 1);
        
        // Recursively construct the left and right subtrees
        TreeNode* leftSubtree = createBalancedTree(inorderTraversal, start, mid - 1);
        TreeNode* rightSubtree = createBalancedTree(inorderTraversal, mid + 1, end);
        
        // Create a new node with the middle element and attach the subtrees
        return new TreeNode(inorderTraversal[mid], leftSubtree, rightSubtree);
    }
    
public:
    TreeNode* balanceBST(TreeNode* root) {
        // Create a vector to store the inorder traversal of the BST
        vector<int> inorderTraversal;
        inorder(root, inorderTraversal);
        int n = inorderTraversal.size();
        
        // Construct and return the balanced tree
        return createBalancedTree(inorderTraversal, 0, n - 1);
    }
};